### PR TITLE
Support case insensitive identifiers in MongoDB

### LINKED ISSUE #1102: MongoDB collection name case-sensitivity
MongoDB Collection is case sensitive: It is totally valid for one database to have two collections: `Hello`, and `hello`, while Presto-MongoDB Connector use SchemaTableName to represent collection name, SchemaTableName itself is not case sensitive, it causes that we can not query mongodb collection which have uppercase letter in its name.

### PR DESCRIPTION
**Pull request by ebyhr**
Supersedes #1915 and fixes #1102

Recently, we've received this request several times.

### CODE REVIEW

File: presto-docs/src/main/sphinx/connector/mongodb.rst [hunk 1]
```diff
@@ -34,24 +34,25 @@ Configuration Properties
...
-``mongodb.ssl.enabled``               Use TLS/SSL for connections to mongod/mongos
-``mongodb.read-preference``           The read preference
-``mongodb.write-concern``             The write concern
-``mongodb.required-replica-set``      The required replica set name
-``mongodb.cursor-batch-size``         The number of elements to return in a batch
-===================================== ==============================================================
+========================================== ==============================================================
+Property Name                              Description
+========================================== ==============================================================
+``mongodb.seeds``                          List of all MongoDB servers
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/mongodb.rst:40]: Please extract refactoring ("Make table wider") to a preparatory commit

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 2]
```diff
@@ -606,6 +611,32 @@ else if (value instanceof Document) {
...
     }
 
+    private String toRemoteSchemaName(String schemaName)
+    {
+        if (!caseInsensitiveNameMatching) {
+            return schemaName;
+        }
+        for (String remoteSchemaName : client.listDatabaseNames()) {
+            if (schemaName.equals(remoteSchemaName.toLowerCase(ENGLISH))) {
+                return remoteSchemaName;
```
Comment (findepi) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:621]: How expensive is this? how often is it called during planing of a query?

(in JDBC we have caching for this, so question, if we need one here)

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 3]
```diff
@@ -606,6 +611,32 @@ else if (value instanceof Document) {
...
+    }
+
+    private String toRemoteTableName(String schemaName, String tableName)
+    {
+        if (!caseInsensitiveNameMatching) {
+            return tableName;
+        }
+        for (String remoteTableName : client.getDatabase(schemaName).listCollectionNames()) {
+            if (tableName.equals(remoteTableName.toLowerCase(ENGLISH))) {
+                return remoteTableName;
```
Comment (findepi) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:634]: How expensive is this? how often is it called during planing of a query?

(in JDBC we have caching for this, so question, if we need one here)

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java (hunk 2 above)
Comment (ebyhr) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:621]: The first attempt is 3 times and subsequent attempts are only once. 

Finding the last schema takes:
* ~70 msec / 10,000 schemas
* ~700 msec / 10,000 schemas

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java (hunk 3 above)
Comment (ebyhr) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:634]: The first attempt is 3 times and subsequent attempts are only once. 

Finding the last table takes:
* ~10 msec / 10,000 tables
* ~10 msec / 10,000 tables

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 4]
```diff
@@ -141,15 +144,18 @@ public void shutdown()
 
     public List<String> getAllSchemas()
     {
-        return ImmutableList.copyOf(client.listDatabaseNames());
+        return ImmutableList.copyOf(client.listDatabaseNames()).stream()
```
Comment (findepi) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:147]: `ImmutableList.copyOf` is redundant

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 5]
```diff
@@ -606,6 +611,32 @@ else if (value instanceof Document) {
         return Optional.ofNullable(typeSignature);
     }
 
+    private String toRemoteSchemaName(String schemaName)
+    {
+        if (!caseInsensitiveNameMatching) {
+            return schemaName;
+        }
+        for (String remoteSchemaName : client.listDatabaseNames()) {
```
Comment (findepi) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:619]: add

```
verify(schemaName.equals(schemaName.toLowerCase(ENGLISH)), "schemaName not in lower-case: %s", schemaName);
```

at the top of the method

Comment (findepi) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:619]: same in the table mapping method

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java (hunk 4 above)
Comment (ebyhr) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:147]: `client.listDatabaseNames()` returns `MongoIterable<String>` (≠`List<String>`).

Comment (findepi) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:147]: you have `ImmutableList.copyOf` and `collect(toImmutableList())`

Comment (ebyhr) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:147]: Replaced with `toList()`. Please let me know if my understanding is wrong.

Comment (sajjoseph) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:147]: Was this change saved? I still see `collect(toImmutableList()) `

Comment (ebyhr) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:147]: @sajjoseph `collect(toImmutableList())` is the expected code.

Comment (sajjoseph) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:147]: Thanks. I am referring to your comment earlier where you said that you replaced with toList(). I don't see that change.

**Pull request merged**